### ISSUE TITLE
Segmentation Fault on .NET 5 build for standalone for ARM

### ISSUE THREAD

**Issue by shadowfox87** [dotnet/runtime#41421]
### Description

This bug is reproduceable on any version of .NET 5 after preview 6. 
1. Build a standalone executable for Linux ARM, and use the produce single file option
2. It will give segmentation fault at the point on your code you instantiate a new TcpListener class. It works fine if you try on windows x64. 

This is the line at where you will see it:
```csharp
_listener = new TcpListener(address, port);
```
If you use the TRIM within the produce single file, it misses some assemblies that are used for webclient (given the same settings from before). This won't work even when trying on windows x64.

### Configuration

* Which version of .NET is the code running on? 
   .NET 5 Preview 8 (can use any version after Preview 6)

* What OS and version, and what distro if applicable? 
   Archlinux latest version

* What is the architecture (x64, x86, ARM, ARM64)? 
   ARM

* Do you know whether it is specific to that configuration? 
  It won't work on ARM but works fine on Windows x64 (but will fail if TRIM is used)

### Regression?

* Did this work in a previous build or release of .NET Core, or from .NET Framework?
Yes it worked on all versions before preview 7.

**Comment by joperezr** [comment]
cc: @eerhardt this sounds like we are missing an annotation on System.Net.Sockets causing us to trim things that we shouldn't be trimming.

**Comment by joperezr** [comment]
if it is indeed a missing annotation, then we probably will want to take this for 5.0

**Comment by karelz** [comment]
@joperezr is it something we (Networking team) should help with?

**Comment by joperezr** [comment]
I don't think so, in dotnet 5 we formed a v-team that has been working on making sure our framework is linker-friendly so we've been adding annotations on the framework and I just want to make sure that this issue doesn't mean that we missed one. We can take a look.

**Comment by joperezr** [comment]
@shadowfox87 are you able to get a dump of that so that we can check the managed callstack to see what is the problem here? I tried reproing this using the Preview 8 bits on a RaspberryPi but didn't have any luck. This is the code that I tried:
```c#
    class Program
    {
        static void Main(string[] args)
        {
            var listener = new TcpListener(IPAddress.Any, 8080);
            Console.WriteLine(listener.ToString());
        }
    }
```

I published that for linux-arm and published it Trimmed and single file. I can validate that the linker is in fact running and trimming some assemblies. The result runs fine for me on my Raspberry Pi. Can you send more info regarding repro steps as well as a full code snippet of what you are doing to reproduce this issue?

**Comment by HKunogi** [comment]
![unknown (1)](https://user-images.githubusercontent.com/1687701/91498243-67d98c80-e8fa-11ea-90f7-c019ca254ee3.png)
![unknown (2)](https://user-images.githubusercontent.com/1687701/91498252-6a3be680-e8fa-11ea-9e3b-9838d7a44668.png)
![unknown](https://user-images.githubusercontent.com/1687701/91498258-6d36d700-e8fa-11ea-8e62-bac3f6bcd22e.png)
![unknown (3)](https://user-images.githubusercontent.com/1687701/91498266-6f009a80-e8fa-11ea-8da3-6b0fbfca8327.png)
If unmark the Produce single file, it stops giving the issue and works fine.

**Comment by shadowfox87** [comment]
@joperezr I can confirm the answer above. If you need any further information, @HKunogi can answer.

**Comment by joperezr** [comment]
Can you confirm which version of the SDK are you using? to figure this out try running `dotnet --list-sdks` and `dotnet --version` as it may be that what you are hitting is already addressed by an annotation that was added later.

**Comment by joperezr** [comment]
Note that Trimming unused assemblies is in preview as noted in the screenshot above, so it is likely for things like this to happen. We will definitely investigate and fix if there is anything to address here but do note that this fix might come after 5.0

**Comment by joperezr** [comment]
cc: @vitek-karas @swaroop-sridhar @agocke 

Even though trimming unused assemblies is in preview for dotnet 5, we may still want to investigate why when selecting to produce a single file this starts causing issues but when not doing so things work just fine. For this reason I'm marking this issue as 5.0 so that we investigate at least that bit now, while the overall trimming scenario for unused assemblies will still be in preview for 5.0.

**Comment by agocke** [comment]
The screenshot says TypeLoadException, but the bug says segfault. I'm not sure what we're looking at.

**Comment by joperezr** [comment]
If I understood correctly, the TypeLoadException only happens when they decide to trim their application given that we are likely missing an annotation somewhere causing the linker to strip out a method that is required at runtime. The segfault problem seems to be the one that happens only when they decide to publish as a single file, but seems to not happen when unchecking that, which is what we probably want to take a look at. @HKunogi to confirm if what I understood is correct.

**Comment by agocke** [comment]
Ah yes I missed the middle screenshot, thanks.

**Comment by HKunogi** [comment]
> If I understood correctly, the TypeLoadException only happens when they decide to trim their application given that we are likely missing an annotation somewhere causing the linker to strip out a method that is required at runtime. The segfault problem seems to be the one that happens only when they decide to publish as a single file, but seems to not happen when unchecking that, which is what we probably want to take a look at. @HKunogi to confirm if what I understood is correct.

Yes, exactly that, about the version, its: 5.0.100-preview.8.20417.9

**Comment by vitek-karas** [comment]
Any kind of repro would help. I tried the trimming problem (TypeLoadException), but no luck - it seems to work fine for me.
I will try linux-arm once I have a machine to test it on.

**Comment by HKunogi** [comment]
> Any kind of repro would help. I tried the trimming problem (TypeLoadException), but no luck - it seems to work fine for me.
> I will try linux-arm once I have a machine to test it on.

When i run it on windows it works perfectly fine, the issue only happens at linux arm.

**Comment by eerhardt** [comment]
@HKunogi - looking at your screenshot, I see you are using `Configuration = Release | x64`

![image](https://user-images.githubusercontent.com/8291187/91773080-aea1dd80-ebab-11ea-9c41-9def174855bf.png)

Is that intentional? Is there a reason your Architecture is set to `x64` when you are trying to publish to an `arm` machine?

**Comment by eerhardt** [comment]
Also - I'm unable to reproduce this on my Raspberry Pi using a simple console app the just creates a new `TcpListener`. @HKunogi, if you point to a complete project that reproduces this error, that would really speed up the investigation.

**Comment by HKunogi** [comment]
> @HKunogi - looking at your screenshot, I see you are using `Configuration = Release | x64`
> 
> ![image](https://user-images.githubusercontent.com/8291187/91773080-aea1dd80-ebab-11ea-9c41-9def174855bf.png)
> 
> Is that intentional? Is there a reason your Architecture is set to `x64` when you are trying to publish to an `arm` machine?

Its just the name of the configuration, it is configured properly, after some more testing i found out that using `<TrimMode>Link</TrimMode>` at the csproj file, fix the assembly not found exception and the segmentation fault. For what i read the .net 5 uses defaultly the assembly-level trimming, but when i set it to that member-level trimming, it simply worked fine. (at least it started up fine, needs to have some runtime tests to make sure its not missing anything and crashing).

**Comment by agocke** [comment]
Dup of https://github.com/dotnet/runtime/issues/41609, should be fixed

**Comment by agocke** [comment]
My mistake, didn't notice this was on arm32, not ARM64.

**Comment by vitek-karas** [comment]
It's possibly going to be fixed by https://github.com/dotnet/runtime/pull/42402.

**Comment by whatevergeek** [comment]
oh well... got #SegmentationFault when running 'dotnet new' using the latest #dotnet 5 linux arm64 sdk ... reverting back to dotnet_v5.0.100-preview.1.20111.5 ... will try again when a new arm64 version is released

**Comment by karelz** [comment]
@vitek-karas what would be next step here?
Where should we route the issue?

**Comment by vitek-karas** [comment]
We would need more information about the crash. I haven't seen this repro anywhere (can't try right now, I don't have ARM64 machine handy).

**Comment by karelz** [comment]
Perhaps closing then? Until there is more info available?

**Comment by vitek-karas** [comment]
Agreed - please open a new issue if there's are still issues with the latest .NET 5 (on any platform).

I also just realized that running `dotnet new` is not related to single-file, so if that crashed it's definitely a different issue than the one described above.

**Comment by whatevergeek** [comment]
what other info you need? I was able to reproduce it on ARM aarch64... i'm running debian on userland...

**Comment by vitek-karas** [comment]
Thanks for more info @whatevergeek :
- What version of .NET you're using (both SDK and runtime - just run `dotnet --info`)
- How did you install it (this is mostly interesting if it goes into some uncommon location)
- What command you run - is it really just `dotnet new console` or something more involved?
- What is the OS (although you already answered that one)

I'll ask somebody to repo it (I don't have ARM64 available at home unfortunately.. yet).

**Comment by whatevergeek** [comment]
hi @vitek-karas ,    
thanks for considering this. here are the info.    
Installation is by tarring arm64.tar.gz from the dotnet [arm64 site](https://dotnet.microsoft.com/download/dotnet/thank-you/sdk-5.0.103-linux-arm64-binaries)    

I've investigated further and got the following observations:

## TLDR:
Latest arm64 package (dotnet-sdk-5.0.103-linux-arm64.tar.gz) doesn’t cause segmentation fault anymore when `dotnet new` is executed. However, build is failing without giving info.

## Exhibit A    
### dotnet-sdk-5.0.103: segmentation fault during `dotnet new` is gone, but build is failing w/o info    
![image](https://user-images.githubusercontent.com/7351683/107652562-ad78ee80-6cbb-11eb-8b34-03df2705263a.png)

![image](https://user-images.githubusercontent.com/7351683/107652582-b49ffc80-6cbb-11eb-9809-26f1a405c53a.png)

## Exhibit B    
### Previous version of .NET 5.0 that was working    
![image](https://user-images.githubusercontent.com/7351683/107652701-d0a39e00-6cbb-11eb-9e5f-76ab61154d69.png)

![image](https://user-images.githubusercontent.com/7351683/107652737-dac59c80-6cbb-11eb-8890-05c9347c2dd2.png)

Perhaps, segmentation fault is already resolved (or might be hiding since build failure is not showing).
Anyway, let me know if you or your team need further info.

I'm really looking forward to use the latest .NET 5.0 on my arm64 device.

**Comment by vitek-karas** [comment]
@whatevergeek if I understand it correctly this is happening on Linux running in userland (as in https://userland.tech/), right? That is a configuration we don't test as far as I know, so it's not surprising it doesn't work. Feel free to create a new issue to request support for this configuration.

**Comment by whatevergeek** [comment]
hi @vitek-karas ,

That's correct. I'm using linux via userland.
Ok will create a new issue to request support for the configuration.
Thanks for the consideration.
